File: 2025/day05.py
intervals = []
ids = []

with open("2025/input/day05") as f:
    data = f.read().splitlines()

    for row in data:
        if "-" in row:
            intervals.append(tuple(int(x) for x in row.split("-")))
        elif row != "":
            ids.append(int(row))


intervals.sort()

valid_id_range = []

for a, b in intervals:
    for i1, i2 in valid_id_range:
        if i1 <= a <= i2:
            if i2 < b:
                valid_id_range.remove((i1, i2))
                valid_id_range.append((i1, b))
            break
    else:
        valid_id_range.append((a, b))

partie1 = sum(any(a <= i <= b for a, b in valid_id_range) for i in ids)

print("Partie 1:", partie1)

partie2 = sum(b - a + 1 for a, b in valid_id_range)

print("Partie 2:", partie2)

File: 2025/rust/src/aoc2025/day05.rs
use rdcl_aoc_helpers::input::WithReadLines;
use std::fs::File;

pub fn day05(path: &String) {
    let data: Vec<String> = File::open(path).read_lines::<String>(1).collect();
    let mut intervals: Vec<(u64, u64)> = Vec::new();
    let mut merged_intervals: Vec<(u64, u64)> = Vec::new();
    let mut ids: Vec<u64> = Vec::new();

    for line in data {
        if line.contains('-') {
            let interval: Vec<u64> = line.split('-').map(|x| x.parse::<u64>().unwrap()).collect();
            intervals.push((interval[0], interval[1]));
        } else if line.len() > 0 {
            ids.push(line.parse::<u64>().unwrap());
        }
    }

    intervals.sort_by(|a, b| a.0.cmp(&b.0));
    merged_intervals.push(intervals[0]);

    for interval in intervals {
        let last = merged_intervals.pop().unwrap();
        if last.0 <= interval.0 && interval.0 <= last.1 {
            if last.1 < interval.1 {
                merged_intervals.push((last.0, interval.1));
            } else {
                merged_intervals.push(last);
            }
        } else {
            merged_intervals.push(last);
            merged_intervals.push(interval);
        }
    }

    let partie1: u64 = ids
        .iter()
        .filter(|x| merged_intervals.iter().any(|y| y.0 <= **x && y.1 >= **x))
        .count() as u64;
    println!("Partie 1: {}", partie1);
    let partie2: u64 = merged_intervals.iter().map(|x| x.1 - x.0 + 1).sum();

    println!("Partie 2: {}", partie2);
}
